openedx/features/content_type_gating/services.py

```diff
@@ -29,4 +46,40 @@ def content_type_gate_for_block(self, user, block, course_id):
             access = has_access(user, 'load', block, course_id)
             if (not access and access.error_code == 'incorrect_user_group'):
                 return access.user_fragment
+
+            # Handle case of staff users masquerading as Learner in Audit or Learner in Limited Access
+            if self._is_masquerading_as_audit_or_limited_access(user, course_id):
+                # Check merged_group_access in case the course steam overrode content type gating
+                # for this problem with the studio override
+                if LIMITED_ACCESS.id not in block.merged_group_access.get(CONTENT_GATING_PARTITION_ID):
+                    return create_content_gating_partition(course).access_denied_fragment(
+                        course, user, LIMITED_ACCESS, [FULL_ACCESS]
+                    )
             return None
+
+    def check_children_for_content_type_gating_paywall(self, item, course, course_id):
+        """
+        Arguments:
+            item (xblock such as a sequence or vertical block)
+            course (xblock.internal.CourseDescriptorWithMixins)
+            course_id (CourseLocator)
+
+        Returns:
+            If:
+            This xblock contains problems which this user cannot load due to content type gating
+            Then:
+            Return the first content type gating paywall (Fragment)
+            Else:
+            Return None
+        """
+        user = self._get_user()
+        if user:
```

Comment:
This is a totally personal nit, please ignore if you disagree even a little bit.

But I generally find functions easier to read if the "happy/common path" is at zero-indentation. Which means I generally do early-exit checks, rather than conditionals like this that indent code for the rest of the function.

Also makes adding or removing conditions less noisy in future diffs.

So like:
```
if not user:
    return None
```

But again, not something I'm religious about. And I don't think I've convinced other members of the team, like Dillon. But just wanted to make the case at least once with ya.  :)